crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -334,163 +334,465 @@ pub fn decompose_dense(
+            }
+            // The middle-levels `stack` is a LIFO, so if we push in this order, we'll consider the
+            // Pauli terms in lexicographical order, which is the canonical order from
+            // `SparsePauliOp.sort`.  Populating the `out_list` (an initially empty `Vec`)
+            // effectively reverses the stack, so we want to push its elements in the IXYZ order.
+            if loc.qubit() == 1 {
+                i_nonzero.then(|| out_list.push(loc.push_i()));
+                x_nonzero.then(|| out_list.push(loc.push_x()));
+                y_nonzero.then(|| out_list.push(loc.push_y()));
+                z_nonzero.then(|| out_list.push(loc.push_z()));
+            } else {
+                z_nonzero.then(|| stack.push(loc.push_z()));
+                y_nonzero.then(|| stack.push(loc.push_y()));
+                x_nonzero.then(|| stack.push(loc.push_x()));
+                i_nonzero.then(|| stack.push(loc.push_i()));
             }
-            phases[i].write(phase % 4);
         }
-        // These are safe because the above loops write into every element.  It's guaranteed that
-        // each of the elements of the `paulis` vec will have `num_qubits` because they're all
-        // reading from the same base array.
-        let z = unsafe { z.assume_init() };
-        let x = unsafe { x.assume_init() };
-        let phases = unsafe { phases.assume_init() };
-        Ok(ZXPaulis {
-            z: z.into_pyarray_bound(py).into(),
-            x: x.into_pyarray_bound(py).into(),
-            phases: phases.into_pyarray_bound(py).into(),
-            coeffs: PyArray1::from_vec_bound(py, coeffs).into(),
-        })
     }
+    (stack, out_list, scratch)
 }
 
-/// Recurse worker routine of `decompose_dense`.  Should be called with at least one qubit.
-fn decompose_dense_inner(
-    factor: Complex64,
+/// Iteratively decompose the matrix at all levels other than the first and last.  This populates
+/// the `out_list` with locations.
+fn decompose_middle_levels(
+    mut stack: Vec<PauliLocation>,
+    out_list: &mut Vec<PauliLocation>,
+    scratch: &mut [Complex64],
     num_qubits: usize,
-    paulis: &[Pauli],
-    block: ArrayView2<Complex64>,
-    out_paulis: &mut Vec<Vec<Pauli>>,
-    out_coeffs: &mut Vec<Complex64>,
-    square_tolerance: f64,
 ) {
-    if num_qubits == 0 {
-        // It would be safe to `return` here, but if it's unreachable then LLVM is allowed to
-        // optimise out this branch entirely in release mode, which is good for a ~2% speedup.
-        unreachable!("should not call this with an empty operator")
-    }
-    // Base recursion case.
-    if num_qubits == 1 {
-        let mut push_if_nonzero = |extra: Pauli, value: Complex64| {
-            if value.norm_sqr() <= square_tolerance {
-                return;
+    let side = 1 << num_qubits;
+    let zero = Complex64::new(0.0, 0.0);
+    // The stack is a LIFO queue, which is how we implement the depth-first iteration.  Depth-first
```

Comment:
```suggestion
    // The stack is a LIFO, which is how we implement the depth-first iteration.  Depth-first
```

what's a "LIFO queue" than a contradiction in terms? lol